File: src/main/java/fi/quanfoxes/parser/patterns/ElsePattern.java
package fi.quanfoxes.parser.patterns;

import java.util.List;

import fi.quanfoxes.Keywords;
import fi.quanfoxes.lexer.ContentToken;
import fi.quanfoxes.lexer.KeywordToken;
import fi.quanfoxes.lexer.ParenthesisType;
import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.lexer.TokenType;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Parser;
import fi.quanfoxes.parser.Pattern;
import fi.quanfoxes.parser.nodes.ElseNode;

public class ElsePattern extends Pattern {
    public static final int PRIORITY = 15;

    private static final int ELSE = 0;
    private static final int BODY = 2;

    public ElsePattern() {
        // Pattern:
        // else [\n] {...}
        super(
                TokenType.KEYWORD, /* else */
                TokenType.END | TokenType.OPTIONAL, /* [\n] */
                TokenType.CONTENT /* {...} */
        );
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        KeywordToken keyword = (KeywordToken)tokens.get(ELSE);

        if (keyword.getKeyword() != Keywords.ELSE) {
            return false;
        }

        ContentToken body = (ContentToken)tokens.get(BODY);
        return body.getParenthesisType() == ParenthesisType.CURLY_BRACKETS;
    }

    private List<Token> getBody(List<Token> tokens) {
        ContentToken body = (ContentToken)tokens.get(BODY);
        return body.getTokens();
    }

	@Override
	public Node build(Context environment, List<Token> tokens) throws Exception {
        Context context = new Context();
        context.link(environment);

        List<Token> body = getBody(tokens);
        Node node = Parser.parse(context, body);

        return new ElseNode(context, node);
	}
}

File: Zigzag/Parser/Patterns/ElsePattern.cs
using System.Collections.Generic;

public class ElsePattern : Pattern
{
	public const int PRIORITY = 1;

	public const int FORMER = 0;
	public const int ELSE = 2;
	public const int BODY = 4;

	// $([else] if) [\n] else [\n] (...)
	public ElsePattern() : base
	(
		TokenType.DYNAMIC, 
		TokenType.END | TokenType.OPTIONAL, 
		TokenType.KEYWORD,
		TokenType.END | TokenType.OPTIONAL, 
		TokenType.CONTENT
	) { }

	public override int GetPriority(List<Token> tokens)
	{
		return PRIORITY;
	}

	public override bool Passes(Context context, PatternState state, List<Token> tokens)
	{
		if (tokens[ELSE].To<KeywordToken>().Keyword != Keywords.ELSE)
		{
			return false;
		}
		
		var former = tokens[FORMER].To<DynamicToken>();

		return former.Node.GetNodeType() == NodeType.IF_NODE ||
				former.Node.GetNodeType() == NodeType.ELSE_IF_NODE;
	}

	public override Node? Build(Context environment, List<Token> tokens)
	{
		var body = tokens[BODY].To<ContentToken>();

		var context = new Context();
		context.Link(environment);

		return new ElseNode(context, Parser.Parse(context, body.GetTokens()));
	}

	public override int GetStart()
	{
		return 1;
	}
}
